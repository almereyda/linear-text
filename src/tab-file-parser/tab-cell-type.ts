import type {TabCell} from '../tab/tab-record'

import {isNumeric} from '../string-util'

/** String-encoded data type of the cell value. */
export type TabCellType =
  /** Nonempty string. */
  | 'text'
  /** Finite number. */
  | 'number'
  /** True or false. */
  | 'bool'
  /** Whitespace-only string. */
  | 'blank'
  /** No value. */
  | 'empty'

/** Infer the string-encoded data type of the cell value. */
export function parseCellType(cell: TabCell): TabCellType {
  if (cell === '') return 'empty'
  if (/^\s+$/.test(cell)) return 'blank'
  const trimmed = cell.trim()
  if (/^(true|false)$/i.test(trimmed)) return 'bool'
  if (isNumeric(trimmed)) return 'number'
  return 'text'
}
