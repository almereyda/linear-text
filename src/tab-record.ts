/** A table row and its application state modeling. */
export type TabRecord = {
  id: number

  /** The modeled text to be shown, possibly mini-Markdown or empty. */
  text: string | undefined

  /**
   * True when the row property is outdated. Invalidated data is flushed to the
   * row on save. Invalidating, saving, and reloading data is expected to
   * produce the same model results as just invalidating and saving.
   */
  invalidated: boolean

  /**
   * The model is flushed to the row only on save. Only row data is written to
   * disk.
   */
  row: TabRow
}

/**
 * Recognized model columns. A given file may label these columns differently.
 */
export type TabColumn = keyof TabRecord & 'text'

/**
 * The ordered row data verbatim, if any. An empty string item indicates no
 * value and undefined indicates the cell was either missing or empty and the
 * last column.
 *
 * This data may be unparseable or unexpected for a given column.
 */
export type TabRow = (string | undefined)[]

/**
 * Model column to row index. A given row may have more or less columns than
 * the header or other rows.
 */
export type TabHeader = Record<TabColumn, number>

let uid: number = 0

export function TabRecord(text?: string | undefined): TabRecord {
  uid++
  return {id: uid, text, invalidated: true, row: []}
}

export namespace TabRecord {
  export function fromRow(row: TabRow, header: TabHeader): TabRecord {
    uid++
    return {id: uid, text: row[header.text], invalidated: false, row}
  }

  export function setText(record: TabRecord, text: string): void {
    record.text = text
    record.invalidated = true
  }
}
