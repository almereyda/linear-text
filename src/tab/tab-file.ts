import type {TabColumnMap, TabRecord, TabRow} from './tab-record'

export type TabFile = Readonly<{meta: TabMeta; records: TabRecord[]}>

export type TabMeta = Readonly<{
  header: TabHeader | undefined
  /** If header is not present, this will be inferred from the data. */
  columnMap: TabColumnMap
  delimiter: string
  newline: string
}>

/**
 * If present, this will be serialized and used to infer column map. This
 * can't currently be modified by the program.
 */
export type TabHeader = TabRow

export function TabFile(): TabFile {
  return {meta: TabMeta(), records: []}
}

export function TabMeta(): TabMeta {
  return {columnMap: {}, ...tabMetaDefaults}
}

export const tabMetaDefaults = Object.freeze({
  header: undefined,
  delimiter: '\t',
  newline: '\n'
})
