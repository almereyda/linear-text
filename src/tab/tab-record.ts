import {IDFactory, makeID} from '../id-factory/id-factory'

/** A table row and its application state modeling. */
export type TabRecord = {
  /**
   * Unique for the program's execution lifetime. Never serialized to row.
   * Symbols are not used as they cannot be serialized for the store.
   */
  readonly id: number

  /** The normalized model text to be shown, possibly mini-Markdown or empty. */
  text: string | undefined

  /**
   * True when the row property is outdated. Invalidated data is flushed to the
   * row on save. Invalidating, saving, and reloading data is expected to
   * produce the same model results as just invalidating and saving.
   */
  invalidated: boolean

  /**
   * The model is flushed to the row only on save. Only row data is written to
   * disk.
   */
  readonly row: TabRow
}

/**
 * Recognized model columns. A given file may label these columns differently.
 */
export type TabColumn = keyof TabRecord & 'text'

/**
 * The ordered row data verbatim, if any. An empty string item indicates no
 * value and undefined indicates the cell was either missing or empty and the
 * last column.
 *
 * This data may be unparseable or unexpected for a given column.
 */
export type TabRow = TabCell[]
export type TabCell = string

/**
 * Model column to row index. A given row may have more or less columns than
 * the header or other rows. The mapping may be sparse.
 */
export type TabColumnMap = Partial<Record<TabColumn, number | undefined>>

export function TabRecord(
  factory: IDFactory,
  text?: string | undefined
): TabRecord {
  return {id: makeID(factory), text, invalidated: true, row: []}
}

// new uninvalidated row
TabRecord.fromRow = (
  factory: IDFactory,
  row: TabRow,
  text: string | undefined
): TabRecord => {
  return {id: makeID(factory), text, invalidated: false, row}
}

TabRecord.setText = (record: TabRecord, text: string): void => {
  record.text = text
  record.invalidated = true
}

// "void"
TabRecord.isSpacing = (record: TabRecord): boolean => {
  return record.text == null || record.text.length === 0
}
