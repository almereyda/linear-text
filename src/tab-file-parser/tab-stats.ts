import type {TabCellType} from './tab-cell-type'
import type {TabColumnMap, TabRow} from '../tab/tab-record'

import {parseCellType} from './tab-cell-type'

export type TabStats = Readonly<{
  // Column index to TabColumnStats.
  columns: TabColumnStats[]
  total: TabColumnStats
}>

export function TabStats(): TabStats {
  return {columns: [], total: TabColumnStats()}
}

/** Tally rows. */
TabStats.sample = (
  stats: TabStats,
  ...rows: readonly Readonly<TabRow>[]
): void => {
  rows.forEach(row => sampleRow(stats, row))
}

function sampleRow({columns, total}: TabStats, row: Readonly<TabRow>): void {
  row.forEach((cell, index) => {
    const type = parseCellType(cell)
    // [todo]: ??=.
    if (columns[index] == null) columns[index] = TabColumnStats()
    columns[index]![type]++
    total[type]++
  })
}

TabStats.inferMap = (stats: TabStats): TabColumnMap => {
  const map: TabColumnMap = {}
  const textIndex = findMaxCellTypeIndex(stats, 'text')
  if (textIndex != null) map.text = textIndex
  return map
}

function findMaxCellTypeIndex(
  stats: TabStats,
  type: TabCellType
): number | undefined {
  if (stats.total[type] === 0) return
  return stats.columns
    .map(columns => columns[type])
    .reduce(
      (maxCountIndex, count, index, types) =>
        count > types[maxCountIndex]! ? index : maxCountIndex,
      0
    )
}

/** Column type statistics. Missing cells within the column are uncounted. */
export type TabColumnStats = Record<TabCellType, number>

export function TabColumnStats(): TabColumnStats {
  return {text: 0, number: 0, bool: 0, blank: 0, empty: 0}
}
