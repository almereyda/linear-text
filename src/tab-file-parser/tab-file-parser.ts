import type {TabFile, TabHeader} from '../tab/tab-file'
import type {TabRow} from '../tab/tab-record'

import {tabMetaDefaults} from '../tab/tab-file'
import Papa from 'papaparse'
import {TabColumnMap, TabRecord} from '../tab/tab-record'
import {TabStats} from './tab-stats'

export function parseTabFile(input: File | string): Promise<TabFile> {
  return new Promise((resolve, reject) =>
    Papa.parse<TabRow>(input, {
      complete({errors, meta, data}) {
        const delimiterErrorIndex = errors.findIndex(
          ({code}) => code === 'UndetectableDelimiter'
        )
        if (delimiterErrorIndex > -1) {
          errors.splice(delimiterErrorIndex, 1)
          meta.delimiter = tabMetaDefaults.delimiter
        }
        if (errors.length > 0) return reject(errors)
        return resolve(parse(meta, data))
      }
    })
  )
}

function parse(parseMeta: Papa.ParseMeta, data: TabRow[]): TabFile {
  const {header, rows} = parseTabHeaderAndRows(data)
  let columnMap: TabColumnMap = {}
  if (header == null) {
    // No header. Use a heuristic to guess the best mapping from the model
    // columns to the actual row columns.
    const stats = TabStats()
    TabStats.sample(stats, ...rows)
    columnMap = TabStats.inferMap(stats)
  } else {
    // Use the header to map from the model columns to the actual row columns.
    columnMap = {text: findHeaderTextIndex(header)}
  }

  return {
    meta: {
      header,
      columnMap,
      delimiter: parseMeta.delimiter,
      newline: parseMeta.linebreak
    },
    records: rows.map(row => parseTabRecord(row, columnMap))
  }
}

function parseTabHeaderAndRows(data: readonly TabRow[]): {
  header: TabHeader | undefined
  rows: readonly TabRow[]
} {
  if (data[0] == null || !isHeader(data[0]))
    return {header: undefined, rows: data}
  const [header, ...rows] = data
  return {header, rows}
}

function parseTabRecord(row: TabRow, columnMap: TabColumnMap): TabRecord {
  const text = columnMap.text == null ? undefined : row[columnMap.text]
  return TabRecord.fromRow(row, text)
}

function isHeader(row: Readonly<TabRow>): boolean {
  return findHeaderTextIndex(row) != null
}

function findHeaderTextIndex(row: Readonly<TabRow>): number | undefined {
  // Find first index of a text column.
  const index = row.findIndex(cell => /^\s*text\s*$/i.test(cell))
  return index === -1 ? undefined : index
}
