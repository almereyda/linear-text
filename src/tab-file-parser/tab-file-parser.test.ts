import {parseTabFile} from './tab-file-parser'
import {readFileSync} from 'fs'
import {TabRecord} from '../tab/tab-record'

test.each(['\t', ','])('delimiter: "%s"', async delimiter => {
  const row = ['a', 'b', 'c'].join(delimiter)
  const expected = {
    meta: {header: undefined, columnMap: {text: 0}, delimiter, newline: '\n'},
    records: [TabRecord.fromRow(['a', 'b', 'c'], 'a')]
  }
  expect(await parseTabFile(row)).toStrictEqual(expected)
})

test.each([
  ['linefeed', '\n'],
  ['carriage return and linefeed', '\r\n']
])('newline: %s', async (_, newline) => {
  const input = `text${newline}abc`
  const expected = {
    meta: {header: ['text'], columnMap: {text: 0}, delimiter: '\t', newline},
    records: [TabRecord.fromRow(['abc'], 'abc')]
  }
  expect(await parseTabFile(input)).toStrictEqual(expected)
})

test('header padding is ignored and preserved', async () => {
  const input = ' animal , text , color \n frog , ribbit , green '
  const expected = {
    meta: {
      header: [' animal ', ' text ', ' color '],
      columnMap: {text: 1},
      delimiter: ',',
      newline: '\n'
    },
    records: [TabRecord.fromRow([' frog ', ' ribbit ', ' green '], ' ribbit ')]
  }
  expect(await parseTabFile(input)).toStrictEqual(expected)
})

test('extra empty columns are ignored and preserved', async () => {
  const input = [
    'a							b							text										d',
    '1							2							3										4'
  ].join('\n') // prettier-ignore
  const expected = {
    meta: {
      header: [
        'a', '', '', '', '', '', '',
        'b', '', '', '', '', '', '',
        'text', '', '', '', '', '', '', '', '', '',
        'd'
      ], // prettier-ignore
      columnMap: {text: 14},
      delimiter: '\t',
      newline: '\n'
    },
    records: [
      TabRecord.fromRow(
        [
          '1', '', '', '', '', '', '',
          '2', '', '', '', '', '', '',
          '3', '', '', '', '', '', '', '', '', '',
          '4'
        ], // prettier-ignore
        '3'
      )
    ]
  }
  expect(await parseTabFile(input)).toStrictEqual(expected)
})

test('header capitalization is ignored and preserved', async () => {
  const input = ' animal , Text , color \n frog , ribbit , green '
  const expected = {
    meta: {
      header: [' animal ', ' Text ', ' color '],
      columnMap: {text: 1},
      delimiter: ',',
      newline: '\n'
    },
    records: [TabRecord.fromRow([' frog ', ' ribbit ', ' green '], ' ribbit ')]
  }
  expect(await parseTabFile(input)).toStrictEqual(expected)
})

test.each([
  ['empty file', '', []],
  [
    'empty file with trailing newline',
    '\n',
    [TabRecord.fromRow([''], undefined), TabRecord.fromRow([''], undefined)]
  ],
  ['empty file with header', 'text\n', [TabRecord.fromRow([''], '')]],
  [
    'nonempty file trailing newline',
    'a\n',
    [TabRecord.fromRow(['a'], 'a'), TabRecord.fromRow([''], '')]
  ],
  [
    'nonempty file with header and trailing newline',
    'text\na\n',
    [TabRecord.fromRow(['a'], 'a'), TabRecord.fromRow([''], '')]
  ]
])('trailing newline is empty record: %s', async (_, input, expected) => {
  expect((await parseTabFile(input)).records).toStrictEqual(expected)
})

test.each([
  ['missing leading cell', ',2,3', TabRecord.fromRow(['', '2', '3'], '2')],
  ['missing middling cell', '1,,3', TabRecord.fromRow(['1', '', '3'], '')],
  ['missing trailing cell', '1,2', TabRecord.fromRow(['1', '2'], '2')],
  [
    'missing trailing cell and trailing comma',
    '1,2,',
    TabRecord.fromRow(['1', '2', ''], '2')
  ],
  ['missing row', '', TabRecord.fromRow([''], undefined)],
  ['extra cell', '1,2,3,4', TabRecord.fromRow(['1', '2', '3', '4'], '2')]
])('missing cells and inconsistencies: %s', async (_, row, expectedRecord) => {
  const header = 'a,text,c\n'
  const input = header + row
  const expected = {
    meta: {
      header: ['a', 'text', 'c'],
      columnMap: {text: 1},
      delimiter: ',',
      newline: '\n'
    },
    records: [expectedRecord]
  }
  expect(await parseTabFile(input)).toStrictEqual(expected)
})

test.each(['games.test.tab', 'groceries.test.tab'])(
  'integration %s',
  async filename => {
    const absoluteFilename = `${__dirname}/tab-file-test-data/${filename}`
    const input = readFileSync(absoluteFilename).toString()
    expect(await parseTabFile(input)).toMatchSnapshot()
  }
)
