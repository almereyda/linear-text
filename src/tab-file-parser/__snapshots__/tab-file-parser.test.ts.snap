// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`integration games.test.tab 1`] = `
Object {
  "meta": Object {
    "columnMap": Object {
      "text": 0,
    },
    "delimiter": "	",
    "header": Array [
      "text",
      "year",
      "wiki",
    ],
    "newline": "
",
  },
  "records": Array [
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "Nintendo Entertainment System",
        "1985",
        "https://wikipedia.org/wiki/Nintendo_Entertainment_System",
      ],
      "text": "Nintendo Entertainment System",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Mega Man 6",
        "1994",
        "https://wikipedia.org/wiki/Mega_Man_6",
      ],
      "text": "- Mega Man 6",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Solar Jetman: Hunt for the Golden Warpship",
        "1990",
        "https://wikipedia.org/wiki/Solar_Jetman",
      ],
      "text": "- Solar Jetman: Hunt for the Golden Warpship",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Super Mario Bros. 3",
        "1989",
        "https://wikipedia.org/wiki/Super_Mario_Bros._3",
      ],
      "text": "- Super Mario Bros. 3",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Teenage Mutant Ninja Turtles III: The Manhattan Project",
        "1992",
        "https://wikipedia.org/wiki/Teenage_Mutant_Ninja_Turtles_III:_The_Manhattan_Project",
      ],
      "text": "- Teenage Mutant Ninja Turtles III: The Manhattan Project",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- The Guardian Legend",
        "1989",
        "https://wikipedia.org/wiki/The_Guardian_Legend",
      ],
      "text": "- The Guardian Legend",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "PC",
      ],
      "text": "PC",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Red Dead Redemption 2",
        "2018",
        "https://wikipedia.org/wiki/Red_Dead_Redemption_2",
      ],
      "text": "- Red Dead Redemption 2",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "PlayStation 3",
        "2006",
        "https://wikipedia.org/wiki/PlayStation_3",
      ],
      "text": "PlayStation 3",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Ico (The Ico & Shadow of the Colossus Collection)",
        "2001",
        "https://wikipedia.org/wiki/The_Ico_%26_Shadow_of_the_Colossus_Collection",
      ],
      "text": "- Ico (The Ico & Shadow of the Colossus Collection)",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Metal Gear Solid 2: Sons of Liberty (Metal Gear Solid HD Collection)",
        "2001",
        "https://wikipedia.org/wiki/Metal_Gear_Solid_HD_Collection",
      ],
      "text": "- Metal Gear Solid 2: Sons of Liberty (Metal Gear Solid HD Collection)",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- Shadow of the Colossus (The Ico & Shadow of the Colossus Collection)",
        "2005",
        "https://wikipedia.org/wiki/The_Ico_%26_Shadow_of_the_Colossus_Collection",
      ],
      "text": "- Shadow of the Colossus (The Ico & Shadow of the Colossus Collection)",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "PlayStation 4",
        "2013",
        "https://wikipedia.org/wiki/PlayStation_4",
      ],
      "text": "PlayStation 4",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- The Last of Us Remastered",
        "2013",
        "https://wikipedia.org/wiki/The_Last_of_Us#The_Last_of_Us_Remastered",
      ],
      "text": "- The Last of Us Remastered",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "Super Nintendo",
        "1991",
        "https://wikipedia.org/wiki/Super_Nintendo_Entertainment_System",
      ],
      "text": "Super Nintendo",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- EarthBound",
        "1995",
        "https://wikipedia.org/wiki/EarthBound",
      ],
      "text": "- EarthBound",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "Switch",
      ],
      "text": "Switch",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "- The Legend of Zelda: Breath of the Wild",
        "2017",
        "https://wikipedia.org/wiki/The_Legend_of_Zelda:_Breath_of_the_Wild",
      ],
      "text": "- The Legend of Zelda: Breath of the Wild",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
  ],
}
`;

exports[`integration groceries.test.tab 1`] = `
Object {
  "meta": Object {
    "columnMap": Object {
      "text": 0,
    },
    "delimiter": "	",
    "header": undefined,
    "newline": "
",
  },
  "records": Array [
    Object {
      "invalidated": false,
      "row": Array [
        "Produce",
      ],
      "text": "Produce",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "[] Cucumber",
      ],
      "text": "[] Cucumber",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "Snacks",
      ],
      "text": "Snacks",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "[x] Pretzels",
      ],
      "text": "[x] Pretzels",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "Frozen",
      ],
      "text": "Frozen",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "[x] Ice Cream",
      ],
      "text": "[x] Ice Cream",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "[x] Pizza",
      ],
      "text": "[x] Pizza",
    },
    Object {
      "invalidated": false,
      "row": Array [
        "",
      ],
      "text": "",
    },
  ],
}
`;
